qiskit/transpiler/passes/routing/star_prerouting.py

```diff
@@ -0,0 +1,409 @@
+
+        Returns:
+            List[StarBlock]: a list of star blocks in the given dag
+            Union[List[DAGOpNode], List[DAGDepNode]]: a list of operations specifying processing order
+        """
+        self.dag = dag
+        blocks, processing_order = self.collect_all_matching_blocks(min_block_size=min_block_size)
+        return blocks, processing_order
+
+    def star_preroute(self, dag, blocks, processing_order):
+        """Returns star blocks in dag and the processing order of nodes within these star blocks
+        Args:
+            dag (DAGCircuit or DAGDependency): a dag on which star blocks should be determined.
+            blocks (List[StarBlock]): a list of star blocks in the given dag.
+            processing_order (Union[List[DAGOpNode], List[DAGDepNode]]): a list of operations specifying
+            processing order
+
+        Returns:
+            new_dag: a dag specifying the pre-routed circuit
+            qubit_mapping: the final qubit mapping after pre-routing
+        """
+        node_to_block_id = {}
+        for i, block in enumerate(blocks):
+            for node in block.get_nodes():
+                node_to_block_id[node] = i
+
+        new_dag = dag.copy_empty_like()
+        processed_block_ids = set()
+        qubit_mapping = list(range(len(dag.qubits)))
+
+        def _apply_mapping(qargs, qubit_mapping, qubits):
+            return tuple(qubits[qubit_mapping[dag.find_bit(qubit).index]] for qubit in qargs)
+
+        is_first_star = True
+        last_2q_gate = [
+            op
+            for op in reversed(processing_order)
+            if ((len(op.qargs) > 1) and (op.name != "barrier"))
+        ]
+        if len(last_2q_gate) > 0:
+            last_2q_gate = last_2q_gate[0]
+        else:
+            last_2q_gate = None
+
+        int_digits = floor(log10(len(processing_order))) + 1
+        processing_order_s = set(processing_order)
+
+        def tie_breaker_key(node):
+            if node in processing_order_s:
+                return "a" + str(processing_order.index(node)).zfill(int(int_digits))
+            else:
+                return node.sort_key
+
+        for node in dag.topological_op_nodes(key=tie_breaker_key):
+            block_id = node_to_block_id.get(node, None)
+            if block_id is not None:
+                if block_id in processed_block_ids:
+                    continue
+
+                processed_block_ids.add(block_id)
+
+                # process the whole block
+                block = blocks[block_id]
+                sequence = block.nodes
+                center_node = block.center
+
+                if len(sequence) == 2:
+                    for inner_node in sequence:
+                        new_dag.apply_operation_back(
+                            inner_node.op,
+                            _apply_mapping(inner_node.qargs, qubit_mapping, dag.qubits),
+                            inner_node.cargs,
+                        )
```

Comment:
We can avoid the extra runtime checking here because we know the input is valid:

```suggestion
                        new_dag.apply_operation_back(
                            inner_node.op,
                            _apply_mapping(inner_node.qargs, qubit_mapping, dag.qubits),
                            inner_node.cargs,
                            check=False,
                        )
```